### PR TITLE
[libxl] Wrap xl failures into special Exception

### PR DESCRIPTION
**Pull request by crogers1**
When our xl calls return failure, we throw from our new Exception
  class.  Due to the monadic context these calls are executed in,
  they are caught and wrapped in the generic "IO Error" defined in
  Errors.hs. The new Exception class allows us to pass along a
  string description of the error so that it can be displayed
  in the UI or logs, depending on where the call originated. This
  should fix the cryptic, non-descriptive "IO Error." message when
  domain operations such as start fail.

  OXT-435

Signed-off-by: Chris Rogers <rogersc@ainfosec.com>

### CODE REVIEW

**Comment by crogers1** [comment]
This is linked to https://github.com/OpenXT/toolstack-data/pull/42

File: xenmgr/XenMgr/Connect/Xl.hs [hunk 1]
```diff
@@ -69,17 +70,25 @@ import System.Process
...
 
 type NotifyHandler = [String] -> Rpc ()
 type Params = [(String, String)]
 
+data XlExceptionClass = OOMException String | XlException String
+    deriving (Typeable)
+instance Exception XlExceptionClass
+
+instance Show XlExceptionClass where
+    show e@(XlException s) = show s
```
Comment (jean-edouard) [inline xenmgr/XenMgr/Connect/Xl.hs]: This function definition is non-exhaustive and doesn't compile:

    | XenMgr/Connect/Xl.hs:84:4:
    |     Warning: Pattern match(es) are non-exhaustive
    |              In the definition of `show': Patterns not matched: OOMException _

Comment (crogers1) [inline xenmgr/XenMgr/Connect/Xl.hs]: Ah. I'll fix that up tomorrow.

**Comment by crogers1** [comment]
Compiler error fixed.  Removed one of the entries in my instance for Show for some reason.

**Comment by rossphilipson** [comment]
Starting another build of all this.

**Comment by rossphilipson** [comment]
While testing we saw an out of memory exception thrown for something that was clearly not an OOM situation. It seems the logic for this is not quite right yet.

**Comment by crogers1** [comment]
It seems that xl return codes are not even close to being unique.  I probably need to go through xl_cmdimpl.c in the libxl tree and put in more descriptive return values.

**Comment by crogers1** [comment]
Going to close these PRs until I can come up with a better solution.

**Pull request closed without merging**